protocol/contracts/templegold/TempleGoldVesting.sol

```diff
@@ -0,0 +1,444 @@
+            if (_schedule.recipient == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+            if (_schedule.start < block.timestamp) { revert CommonEventsAndErrors.InvalidParam(); }
+            if (_schedule.cliff <= _schedule.start) { revert CommonEventsAndErrors.InvalidParam(); }
+            // total duration must be greater than cliff
+            // duration must be more than difference from start to cliff
+            if (_schedule.duration <= _schedule.cliff - _schedule.start) { revert CommonEventsAndErrors.InvalidParam(); }
+            if (_schedule.amount == 0) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+           
+            bytes32 vestingId = computeNextVestingScheduleIdForHolder(_schedule.recipient);
+            schedules[vestingId] = _schedule;
+            totalVested += _schedule.amount;
+            /// @dev duplicate vesting IDs will not be generated from `computeNextVestingScheduleIdForHolder` so ignoring returned value from `add`
+            _vestingIds.add(vestingId);
+            _activeVestingIds.add(vestingId);
+            _holdersVestingCount[_schedule.recipient] += 1;
+            emit ScheduleCreated(vestingId, _schedule.recipient, _schedule.start, _schedule.cliff, _schedule.duration, _schedule.amount);
+        }
+    }
+
+    /**
+     * @notice Change recipient
+     * @param _vestingId Vesting Id.
+     * @param _recipient Recipient address
+     */
+    function changeRecipient(bytes32 _vestingId, address _recipient) external override onlyElevatedAccess {
+        if (!_vestingIdExists(_vestingId)) { revert CommonEventsAndErrors.InvalidParam(); }
+        if (_recipient == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+
+        VestingSchedule storage _schedule = schedules[_vestingId];
+        address _oldRecipient = _schedule.recipient;
+        if (_oldRecipient == _recipient) { revert CommonEventsAndErrors.InvalidParam(); }
+        _schedule.recipient = _recipient;
+        // create new id
+        bytes32 _newVestingId = computeNextVestingScheduleIdForHolder(_recipient);
+        schedules[_newVestingId] = _schedule;
+        delete schedules[_vestingId];
+
+        _vestingIds.remove(_vestingId);
+        _activeVestingIds.remove(_vestingId);
+        _vestingIds.add(_newVestingId);
+        _activeVestingIds.add(_newVestingId);
+        _holdersVestingCount[_oldRecipient] -= 1;
+        _holdersVestingCount[_recipient] += 1;
+        emit RecipientChanged(_newVestingId, _oldRecipient, _recipient);
+    }
+
+    /**
+     * @notice Revoke vesting
+     * @param _vestingId Vesting Id.
+     * @param _revokeTimestamp Time to revoke
+     */
+    function revokeVesting(bytes32 _vestingId, uint32 _revokeTimestamp) external override onlyElevatedAccess {
+        if (_revokeTimestamp < block.timestamp) { revert CommonEventsAndErrors.InvalidParam(); }
+        VestingSchedule storage _schedule = schedules[_vestingId];
```

Comment:
In this case, because you're accessing all the fields (and multiple times when passing into `_calculateTotalVestedAt` as `memory` anyway), it probably makes sense to use `memory` here. But will need to test the gas utilisation.

Reply:
I use `storage` because later inthe function `_schedule.revoked=true` is written to storage.